Add Multiple Recipients option to the Send form

johnny9 opened the pull request:
This change introduces the multiple recipients controls to the Send form. It is enabled in the ellipses option menu by toggling on "Enable Multiple Recipients". This is stored as a QSetting for the user.

This PR depends on #448 which contains the

![Screenshot from 2025-05-08 11-51-34](https://github.com/user-attachments/assets/c1756abd-b485-4cf1-b6a8-2c0fa91ff05a)
 first implementation of the Send options menu.
![Screenshot from 2025-05-08 11-51-26](https://github.com/user-attachments/assets/6ce9dd39-1b12-49dd-af50-2724fabbb8b9)

johnny9 committed:
qml: Introduce SendRecipientsListModel

The SendRecipientsListModel is owned by WalletQmlModel

johnny9 committed:
qml: Add Multiple Recipients toggle to Send menu

johnny9 committed:
qml: Add Multiple Recipients bar to Send form

johnny9 committed:
qml: Reduce size of recipient selectors

johnny9 committed:
qml: Add remove button to multiple recipients

johnny9 committed:
qml: Prepare transaction with recipients list

johnny9 committed:
qml: Add MultipleSendReview page

MarnixCroes reviewed (commented):
b4e421e19dbbf564431a57319e88630edffb820b

- the amount is not tied to each recipient. If I enter x at 1, all have amount x
- small nit, could this be centered
 
![Screenshot from 2025-05-30 13-26-54](https://github.com/user-attachments/assets/58ffb619-afb6-487f-829d-3d3ec7b32db3)

johnny9 committed:
qml: Add plus big filled icon

johnny9 committed:
qml: Replace NavButton with IconButton in Send

johnny9 committed:
qml: Cleanup BitcoinAmount

This merges BitcoinAmount with SendRecipient to simplify the qml
logic. By doing so, the conversions can be managed all in c++
against the satoshi member variable. Each recipient having its
own BitcoinAmount allows the amounts to be saved independantly
when there are multiple recipients.

johnny9 commented:
> [b4e421e](https://github.com/bitcoin-core/gui-qml/commit/b4e421e19dbbf564431a57319e88630edffb820b)
> 
>     * the amount is not tied to each recipient. If I enter x at 1, all have amount x
> 
>     * small nit, could this be centered
> 
> 
> ![Screenshot from 2025-05-30 13-26-54](https://private-user-images.githubusercontent.com/93143998/449370037-58ffb619-afb6-487f-829d-3d3ec7b32db3.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDg2NzA5MjgsIm5iZiI6MTc0ODY3MDYyOCwicGF0aCI6Ii85MzE0Mzk5OC80NDkzNzAwMzctNThmZmI2MTktYWZiNi00ODdmLTgyOWQtM2QzZWM3YjMyZGIzLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA1MzElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNTMxVDA1NTAyOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTMwZjZkMzE5YzE2MjA5YTJlNTNkNmU2YmFiZmZmYWNmNjU3ZWJkN2E1ZmFjOGYyNDIxOTUyYTQzYjM5MDdiMjYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.VkI2sLuvgvMpOCcBUT1Rxkf67lOL2D2cFpNfkA35E1U)

I believe I have both of these issues resolved now with the latest commits.

johnny9 commented:
Discovered an issue with the calculation of the total amount in the review ui and coin selection ui. Will resolve that today.

johnny9 committed:
qml: Add total calculation to SendRecipientsListModel

johnny9 committed:
qml: Commit Recipient amount when active focus is lost

johnny9 committed:
qml: Clear Send form after sending transaciton

GBKS commented:
Looks great.

1. There seems to be a small issue with the removal logic.

- For recipients 1, 2, and 3, I respectively enter `a`, `b` and `c` into the `Send to` field
- With recipient 1 selected, I click `-` to remove the current recipient (as expected)
- A recipient is removed, I can tell from the header changing from `Recipient 1 of 3` to `Recipient 1 of 2`
- However, I still see `a` in the `Send to` field. I expected the current recipient to be removed and that I now would see `b`.
- I click `>` and now see `c` in the `Send to` field
- I click `<` and now see `b` in the `Send to` field

Not sure if that outline is easy to follow, but by guess is that the current recipient is correctly being removed, but the UI is not being refreshed.

2. There are some visual tweaks to be made, but those can be done separately.

3. Is there a limit on how many recipients can be added? Should we restrict it to something like 25 for now? Michael and I were wondering about that.

4. Also, should the `Note to self` be per recipient or for the whole transaction? In the prototype I have it per recipient (essentially an address label). But in this implementation it's the same for all recipients.

johnny9 committed:
qml: Handle removal of first recipient properly

johnny9 committed:
qml: Restrict Recipients to 25

johnny9 commented:
Update from [e5cd1a7](https://github.com/bitcoin-core/gui-qml/commit/e5cd1a7c23cabd66db18e7ce99bf4dfeb6ee593e) to [8a6b593](https://github.com/bitcoin-core/gui-qml/commit/8a6b593e54673c59c84c43fbdf642da81ce91e64)
 - rebased with main

johnny9 commented:
> Looks great.
> 
>     1. There seems to be a small issue with the removal logic.
>     3. Is there a limit on how many recipients can be added? Should we restrict it to something like 25 for now? Michael and I were wondering about that.
> 

These two issues are now resolved

johnny9 committed:
qml: Split amount and display updating for Bitcoin amount input

MarnixCroes reviewed (commented):
8a6b593e54673c59c84c43fbdf642da81ce91e64
_I found multiple things, probably some are to be fixed and some can be follow-up._

- Have 2 recipients, delete recipient 2. It stays at recipient 2 and displays `Recipient 2 of 1`:

![Screenshot from 2025-06-04 09-30-31](https://github.com/user-attachments/assets/762c03f1-0afb-459c-9aa8-a75aaf2ec808)

- I noticed a delay in the coin selection `Remaining to select` / `Over required amount`, this does not happen on master

[Screencast from 2025-06-04 09-12-47.webm](https://github.com/user-attachments/assets/c5918670-e52f-47a0-9286-e63361b68b74)

- Have multiple recipients, one in sats and the other in BTC -> the Transaction Details screen is confusing as it doesn't have units:

![Screenshot from 2025-06-04 09-14-57](https://github.com/user-attachments/assets/db43a3d5-4335-4341-bcea-29ccd135cf86)


**Some issues when having 2 or more recipients and disabling `Multiple Recipients` as the other disabled recipients are still taken into account:**
- After disabling multiple recipients, the Send view will show the current selected single recipient, but the amount in `Transaction Details` is the sum of the multiple recipients (even though disabled)

[Screencast from 2025-06-04 09-07-22.webm](https://github.com/user-attachments/assets/b4552376-7f8a-4b7d-b854-18228bb98e18)

- Recipient 2 is empty, disable `Multiple Recipients`. Send view is at recipient 1 (with address and amount) but `Review` doesn't work.

rabbitholiness commented:
I also noticed a few additional things to the ones mentioned above:

-  If I switch to Sats as a unit for the amount and then add a new recipient, the unit for that newly added recipient is changed to ₿. I would have expected it to also use Sats.
- Since the address is on one line, it gets cut off. I guess the[ multi-line view](https://lively-kashata-cfde7e.netlify.app/screen/send) just isn't implemented yet? 
<img width="955" alt="Screenshot 2025-06-05 at 9 08 43" src="https://github.com/user-attachments/assets/ea26268c-c1f8-40b0-b37d-b429edf5c5d0" />

- If I use the same address for different recipients, then we should probably highlight that somehow. Based on how difficult it is to implement, there is a more generic way to just say "You're reusing this address for multiple recipients." But ideally we would say "This address is being reused for Recipient x and Recipient y." The names of the recipients could be a link to the corresponding form. Here's how that could look like. 
![address reuse multiple recipients](https://github.com/user-attachments/assets/97e07352-1922-4c4e-82ab-18adde0ea9df)
- I can't get to the review screen, for some reason.
- I think the note to self should be per recipient. Especially since it should be used to identify the payments on the review screen (and on the transaction details page), eventually.

johnny9 committed:
qml: When multipleRecipients is disabled, clearToFront of the list

johnny9 commented:
> [8a6b593](https://github.com/bitcoin-core/gui-qml/commit/8a6b593e54673c59c84c43fbdf642da81ce91e64) _I found multiple things, probably some are to be fixed and some can be follow-up._
> 
>     * Have 2 recipients, delete recipient 2. It stays at recipient 2 and displays `Recipient 2 of 1`:

Found the off by one error. It should work correctly now.

> 
> ![Screenshot from 2025-06-04 09-30-31](https://private-user-images.githubusercontent.com/93143998/451448428-762c03f1-0afb-459c-9aa8-a75aaf2ec808.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDkxODA4NjYsIm5iZiI6MTc0OTE4MDU2NiwicGF0aCI6Ii85MzE0Mzk5OC80NTE0NDg0MjgtNzYyYzAzZjEtMGFmYi00NTljLTlhYTgtYTc1YWFmMmVjODA4LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA2MDYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNjA2VDAzMjkyNlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTExOWEwZWE5NDdhMDk2ODlhMDQ4ZWFmOTA4Zjk2YjBiZDgxYTdkNzMyN2Q1MWE4NWY1NGE3ZTc2YmY1ZmVjYWMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.avrJl8mgvW3a1GNhdvcuqzebpuBKOizDd4zXWJa-tVw)
> 
>     * I noticed a delay in the coin selection `Remaining to select` / `Over required amount`, this does not happen on master

This is a similar fix to the validation. I've push the fix to this PR as well.

> 
> Screencast.from.2025-06-04.09-12-47.webm
> 
>     * Have multiple recipients, one in sats and the other in BTC -> the Transaction Details screen is confusing as it doesn't have units:

The transaction details I will create a new task/issue as a follow up as I would like to do a bunch of amount labels at once.

> 
> ![Screenshot from 2025-06-04 09-14-57](https://private-user-images.githubusercontent.com/93143998/451449690-db43a3d5-4335-4341-bcea-29ccd135cf86.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDkxODA4NjYsIm5iZiI6MTc0OTE4MDU2NiwicGF0aCI6Ii85MzE0Mzk5OC80NTE0NDk2OTAtZGI0M2EzZDUtNDMzNS00MzQxLWJjZWEtMjljY2QxMzVjZjg2LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA2MDYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNjA2VDAzMjkyNlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWNiYjZkMDRlYTdmNzI1NzZhMzQxNDM0MmRlY2U2OTA4MmQ1OTFkYjY5ZTkzNmE3ZTkxYTkxYjRjMDAwMzA5NTcmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.FgU6TrPzmFhr1NHrrO4foCMSKpH2DPQd673YVydyChM)
> 
> **Some issues when having 2 or more recipients and disabling `Multiple Recipients` as the other disabled recipients are still taken into account:**
> 
>     * After disabling multiple recipients, the Send view will show the current selected single recipient, but the amount in `Transaction Details` is the sum of the multiple recipients (even though disabled)
> 
> 
> Screencast.from.2025-06-04.09-07-22.webm
> 
>     * Recipient 2 is empty, disable `Multiple Recipients`. Send view is at recipient 1 (with address and amount) but `Review` doesn't work.

I now have it clearing out the list up to the first recipient when the setting is disabled and the amounts should be correct now.

johnny9 committed:
qml: New recipients use current's units

johnny9 committed:
qml: Connect Recipient label to Note to self input

GBKS commented:
Awesome progress. Not sure if it's covered elsewhere, here are some things I am seeing based on the latest code:

- I can enter infinite characters in the address, amount and note fields. Super long amounts end up overlapping the unit toggle and going outside of the form boundaries.
- I can enter letter (abc...) in the amount input, should be just numbers and comma/period.
- When enabling multiple recipients, it should automatically add a second one.
- In the prototype, I set up an application-wide setting for the unit format. So if you toggle it in the form for one recipient, it also toggles it for all other recipients, the balance in the top left, any other screen you visit, etc. As a mental model, that seems easier track than having individual components have their own setting.
- The disabled state for the pagination/add/remove buttons is super dark, it should be Neutral 4 instead of Neutral 2.
- When a transaction was sent, the `Multiple recipients` setting should also reset to disabled. Weird to be on the screen and having it read `Recipient 1 of 1`.

johnny9 commented:
> Awesome progress. Not sure if it's covered elsewhere, here are some things I am seeing based on the latest code:
> 
>     * I can enter infinite characters in the address, amount and note fields. Super long amounts end up overlapping the unit toggle and going outside of the form boundaries.
> 
>     * I can enter letter (abc...) in the amount input, should be just numbers and comma/period.

These two are addressed in #462 

>     * When enabling multiple recipients, it should automatically add a second one.

This makes sense, will add
>     * In the prototype, I set up an application-wide setting for the unit format. So if you toggle it in the form for one recipient, it also toggles it for all other recipients, the balance in the top left, any other screen you visit, etc. As a mental model, that seems easier track than having individual components have their own setting.

I'll make an issue to track this for a follow up. I need to refactor some things to get the setting to be global.

>     * The disabled state for the pagination/add/remove buttons is super dark, it should be Neutral 4 instead of Neutral 2.

 Will fix.

>     * When a transaction was sent, the `Multiple recipients` setting should also reset to disabled. Weird to be on the screen and having it read `Recipient 1 of 1`.

Ill try to fix this as well.

johnny9 committed:
qml: Add a recipient when multiple recipients are enabled

johnny9 committed:
qml: Disable multipleRecipients after clearing the recipients list

johnny9 committed:
qml: Change IconButton disabled color to neutral4

MarnixCroes commented on src/qml/pages/wallet/MultipleSendReview.qml:
```diff
@@ -0,0 +1,155 @@
+            }
+
+            ContinueButton {
+                id: confimationButton
```
```suggestion
                id: confirmationButton
```

MarnixCroes reviewed (approved):
tested ACK 564231917b43bc0ab121efd90e7a70ec8761faef
LGTM for initial feature introduction

johnny9 committed:
Fix typo in MultipleSendReview

Co-authored-by: Marnix Croes <93143998+MarnixCroes@users.noreply.github.com>
